lua/mason-lspconfig/ensure_installed.lua

```diff
@@ -16,6 +19,8 @@ local function resolve_package(lspconfig_server_name)
 end
 
 return function()
```

Comment:
We could avoid a lot of the state-tracking code by running this in an async context. We'd wrap the entire function with `return a.scope(function () ...  end)` and then we can leverage the `a.wait_all` utility to block execution until all provided "promises" are resolved.

Here's a slimmed down example:

```lua
local a = require "mason-core.async"

---@param pkg Package
local function async_install(pkg)
    ---@async
    return function ()
        return a.wait(function (resolve)
            pkg:install({ version = version } ):once("closed", function ()
                resolve(pkg:is_installed())
            end)
        end)
    end
end

-- a.scope creates a new coroutine scope, allowing us to execute async functions in it
return a.scope(function ()
    local installations = {}
    for _, pkg in ipairs(packages) do
        table.insert(installations, async_install(pkg))
    end

    if not _.all_pass(_.identity, a.wait_all(installations)) then
        -- one or more packages failed
    else
        -- success very nice
    end
end)
```

Reply:
I think it'd be best to keep the amount of events low. Is it enough with simply one event when installation completes? We can pass data to the event containing which packages were installed - then it's up to the consumer to do with that information whatever they want (e.g. check for failures)

Reply:
Ah all of this is great! I also agree that having one event is ideal in this case and totally meets the user case. Thanks for the guidance. I'll update my PR on Monday!